jaxopt/_src/scipy_wrappers.py

```diff
@@ -312,10 +313,19 @@ def scipy_fun(x_onp: onp.ndarray) -> Tuple[onp.ndarray, onp.ndarray]:
                                 options=self.options)
 
     params = tree_util.tree_map(jnp.asarray, onp_to_jnp(res.x))
+    if hasattr(res, 'hess_inv'):
+        if isinstance(res.hess_inv, onp.ndarray):
+            hess_inv = res.hess_inv
+        else:
+            hess_inv = res.hess_inv.todense()
```

Comment:
Silently densifying sparse matrices is not ideal. This could blow up memory... It may be better to raise an exception for the time being.

Reply:
JAX has an experimental sparse module https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html

Reply:
Indeed, I understand why it's a problem to always densify it.
However, most of the time (at least in the test routines), the returned object is not a numpy array, but a `LbfgsInvHessProduct` instance (see [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.LbfgsInvHessProduct.html))... and I could not find a way to extract the matrix as sparse matrix.
Maybe just return the `LbfgsInvHessProduct` as it is? There is no JAX version of it, if I'm correct. I tried to get the hessian out of the operator without turning it into a dense matrix first (e.g. via a product with JAX's sparse matrix, but it returns a numpy dense array). 
Do you have a specific idea of how to use the sparse module from JAX in this context?

Reply:
Sorry for the late reply. From what you said, simply adding the `LbfgsInvHessProduct` object (without converting it to an array) to the state seems like the way to go. What do you think?

Reply:
@mblondel The problem is that this object cannot be cast by default to a JAX object. If I directly return the object itself, I get this error after running the optimizer `'L-BFGS-B'`:
```
TypeError: Value <2x2 LbfgsInvHessProduct with dtype=float64> with type <class 'scipy.optimize.lbfgsb.LbfgsInvHessProduct'> is not a valid JAX type
```
There should be a way to extract the matrix while keeping it as sparse in memory, but I can't figure out how. Or to wrap the scipy's `LbfgsInvHessProduct` object to a JAX-friendly one, but I am not sure how to do this either.

For `'BFGS'` there is no such problem, since the returned hessian is an actual matrix.

Reply:
Or what about allowing the user to decide if the hess_inv should be returned or not? By default it would not be, but if the user knows that memory won't be a problem or if the number of parameters is low, it densifies the matrix and returns it.

Reply:
I feel it should be possible to wrap `LbfgsInvHessProduct` into a pytree (see `RegisteredSpecial2` example [here](https://jax.readthedocs.io/en/latest/pytrees.html))

@froystig Any recommendation on this?

Reply:
Resources on custom pytrees registration:
* https://jax.readthedocs.io/en/latest/pytrees.html
* https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html